### ISSUE TITLE
Player search algorithm needs improvement

### ISSUE THREAD

**Issue by prsauer** [thekroko/uthgard-herald#38]
It looks like
https://uthgard.org/herald/api/search/player/Arms
is using simple prefix matches to return results.

Would be nice to have something a little smarter -- I know @thekroko had suggested a trie and I think that would work pretty well. I think etas in irc had some other suggestion.

**Comment by OliverRadini** [comment]
I agree that something better would be nice, for now we could for sure work with what we have, improved search results wouldn't need any change on the frontend

**Comment by burfo** [comment]
What was suggested?  _trie_?

**Comment by Oricul** [comment]
This should also extend to the guild search.

Essentially, https://uthgard.org/herald/api/search/guild/Dragon should return any guild name containing 'Dragon'. You can compare with https://uthgard.org/herald/api/search/guild/Sword and see that 'Sword of the Dragon' should be returned with the 'Dragon' search.

**Comment by andir** [comment]
On 15:22 16.02.17, Oricul wrote:
> This should also extend to the guild search.
> 
> Essentially, https://uthgard.org/herald/api/search/guild/Dragon should return any guild name containing 'Dragon'. You can compare with https://uthgard.org/herald/api/search/guild/Sword and see that 'Sword of the Dragon' should be returned with the 'Dragon' search.
> 

If the code currently running on the API server is based of what I sent
to @thekroko then it is just prefix search through a radix tree.

Search for random substrings is probably hard to implement. It would
require a full text search through all guild names. Currently that might
not be an issue but lets say ~5 years into the server there will be a
lot more guilds - and a lot of them will be dead or whatever.


One could probably feed the tree with more keys and reference to the
same keys. I'm pretty sure @thekroko is aware of that.

My naive approach would be to feed in guild on word boundaries
(excluding common words e.g. `in`, `of`, and so on).

The guild in your example (`Sword of the Dragon`) would the be searchable
by the phrases `Dragon` and `Sword`.

The main concern is probably about the resource usage on those searches.
Radix trees are pretty good with runtime complexity but it is still a
cost that has to be satisfied.
